[BOLT][AArch64] Handle OpNegateRAState to enable optimizing binaries with pac-ret hardening

bgergely0 opened the pull request:
This PR contains a new approach to replace https://github.com/llvm/llvm-project/pull/117578.

## Problem
Currently BOLT does not support Aarch64 binaries with pointer authentication, because it cannot correctly handle `.cfi_negate_ra_state`.

This has been raised in several issues:

- https://github.com/llvm/llvm-project/issues/74833
- https://github.com/llvm/llvm-project/issues/80992

## Where are OpNegateRAState CFIs needed?

These CFI instructions need to be placed in locations, where one instruction has a different return address state (signed or not signed) than the previous one.

As instructions are moving around during optimization, we cannot know where these should be placed, until all optimizations are done.

For this reason, my implementation does not store these CFIs on the function during binary reading, instead it stores information about them on instructions as MCAnnotation, to be able to work out the states of instructions later.

## MarkRAStates Pass
- this pass processes the extra MCAnnotations to work out the state of each instruction before the optimizations happen
- because the OpNegateRAState only flips a bit, we don't know the RA state before and after it, we only know that it changes
- RememberState and RestoreState also modify the RA state: at restore we have to know the remembered state, so these are also saved as MCAnnotations, and used during the MarkRAStates Pass.

## InsertNegateRAStatePass
- this pass is similar to what I had in the previus version: it looks for changes in the RA state, and adds the CFI to these locations
- the pass is added after other optimizations are done
- the pass has to work out the state of instructions added during optimizations: these instructions inherit the state of the last instruction with known state

## Difference from #117578 
- Previous solution depended on a correct CFG, which is not available (see https://github.com/llvm/llvm-project/issues/115154)
- Instead of discovering the RA State of BasicBlocks using the CFG, this change tracks the state of each individual instruction

## Testing
This PR was tested locally on several workloads, however lit tests are not yet added for all covered cases. As the negate-ra-state CFIs are read by the unwinder, the following situations need to be tested:
- unwinding from C++ exceptions,
- creating stack traces.

## TODOs
- [x] lit test to check negate-ra-state locations without reordering
- [x] lit test to check handling inconsistent inputs in MarkRAStates
- [x] lit test to check `--disallow-pacret` flag
- [x] lit test to check C++ exception unwinding works
- [x] lit test to check BB reordering
- [x] lit test to check function splitting
- [x] write an RFC post to discourse describing the design of the patch
- [x] add a markdown doc to the patch with similar content

## Will be covered in follow-up PRs:
- support for different function splitting strategies
- lite mode
- use-old-text mode

bgergely0 committed:
[BOLT] Recognize paciasp and autiasp instructions

bgergely0 committed:
[BOLT] Support for OpNegateRAState

- save OpNegateRAState, RememberState and RestoreState locations
  when parsing input
- determine the RA states from these before other optimizations
  in MarkRAStates Pass
- after optimizations, we can insert OpNegateRAState at state
  boundaries and other needed locations (e.g. split functions)
  in InsertNegateRAStatePass

kbeyls reviewed (commented):
Thank you very much for this patch @bgergely0 !

I haven't reviewed in depth (and might not be the best person to do an in-depth review), but I thought I'd share my initial thoughts after seeing this patch.

I think the PR description is excellent, it explains the overall design of how to correctly create/transform `.cfi_negate_ra_state` well.
I think it would be best that that description would live not only in the commit message, but ideally in some "design doc" area for BOLT. I don't think there's a "design doc" area for BOLT at the moment (@maksfb , is that right?), so probably best to try and find an appropriate place in the source code to store the info as a comment?

My one other high-level observation is that I'm wondering if it isn't too costly somehow to have many different kinds of MCAnnotation for this one use case. I wonder if it would be better/cheaper to have only a single annotation that would represent "OpNegateRAStateInfo", rather than having separate annotations. (Do I guess correctly that with this patch, typically, there will be a few annotations added to each MCInst?). Maybe some of the long-term BOLT developer might have an informed opinion on that?

bgergely0 commented:
Thanks for the feedback @kbeyls!

## Doc
I agree that storing an in-depth description would be nice. I added comments to key parts, so the sources contain a more concise description  of the problem/solution, but these are scattered in different files.

## MCAnnotations
I'm not sure how a single annotation would work. As the MCAnnotation is essentially an `enum`, I cannot represent the different states by one enum element.

As far as I know (and backed up by https://godbolt.org/z/fv87j3vod ) even though the 4 element enum could be described in 2 bits, a `color` option takes up 4 bytes. So I think we have plenty of space for adding new MCAnnotation options.

The runtime memory footprint would increase a bit, because in MarkRAState Pass, the state of each instruction is annotated, so an extra annotation is stored on each MCInst, but according to my bytehound measurements, it increases the memory by less than 3%. And this only applies in the case of `pac-ret` binaries, which were not supported before, so the comparison is not very meaningful. And because the size of an MCAnnotation does not increase, users running BOLT on non `pac-ret` binaries would not see a difference in runtime memory consumption.

Thanks for pointing this out though, I just realized I left the `kUnknownSign` in, which is not used at all.

paschalis-mpeis commented on bolt/include/bolt/Core/BinaryFunction.h:
```diff
@@ -1599,6 +1599,52 @@ class BinaryFunction {
+      Offset = I->first;
+    }
+    assert(I->first == Offset && "CFI pointing to unknown instruction");
+    if (I == Instructions.begin()) {
+      return {};
+    }
```
nit: could drop braces: 
```suggestion
    if (I == Instructions.begin())
      return {};
```

paschalis-mpeis commented on bolt/include/bolt/Passes/InsertNegateRAStatePass.h:
```diff
@@ -0,0 +1,25 @@
+#ifndef BOLT_PASSES_INSERT_NEGATE_RA_STATE_PASS
```
Please add license notes to this header. Same with `MarkRAStates.h`.

paschalis-mpeis commented on bolt/include/bolt/Passes/InsertNegateRAStatePass.h:
```diff
@@ -0,0 +1,25 @@
+  /// Pass entry point
+  Error runOnFunctions(BinaryContext &BC) override;
+  void runOnFunction(BinaryFunction &BF);
+  bool addNegateRAStateAfterPacOrAuth(BinaryFunction &BF);
```
Please consider making the below 2 methods private.

You may also add a very brief description, and on `addNegateRAStateAfterPacOrAuth`
explain the `\return`.

paschalis-mpeis commented on bolt/include/bolt/Core/MCPlusBuilder.h:
```diff
@@ -1122,6 +1143,51 @@ class MCPlusBuilder {
   /// Return true if the instruction is a tail call.
   bool isTailCall(const MCInst &Inst) const;
 
+  /// Stores NegateRAState annotation on Inst.
```
nit: could use `\p` to indicate parameter (same for the new functions below this)

```suggestion
  /// Stores NegateRAState annotation on \p Inst.
```

paschalis-mpeis commented on bolt/lib/Core/Exceptions.cpp:
```diff
@@ -632,8 +643,13 @@ bool CFIReaderWriter::fillCFIInfoFor(BinaryFunction &Function) const {
       // DW_CFA_GNU_window_save and DW_CFA_GNU_NegateRAState just use the same
```
I think the author made a typo here. Could you fix that in your patch as well?
(I'm asking this as I have related suggestions elsewhere in my review)

```suggestion
      // DW_CFA_GNU_window_save and DW_CFA_AARCH64_negate_ra_state just use the same
```

paschalis-mpeis commented on bolt/lib/Core/Exceptions.cpp:
```diff
@@ -632,8 +643,13 @@ bool CFIReaderWriter::fillCFIInfoFor(BinaryFunction &Function) const {
       if (Function.getBinaryContext().isAArch64()) {
-        Function.addCFIInstruction(
-            Offset, MCCFIInstruction::createNegateRAState(nullptr));
+        // Not adding OpNegateRAState since the location they are needed
+        // depends on the order of BasicBlocks, which changes during
+        // optimizations. Instead, an annotation is added to the instruction, to
+        // mark that the instruction modifies the RA State. The actual state for
+        // instructions are worked out in MarkRAStates based on these
+        // annotations.
```
The first sentence below is based on the fact that previously `addCFIInstruction` was used with OpNegateRAState, right? This does make sense, but for future readers lacking context might be confusing as the relevant code is dropped.

If that is still fundamental info you could reword this. Alternatively you can drop the first sentence, maybe add a note on the commit message if you like, and here simply say something like:
```
// Annotate instruction to mark ...
```

bgergely0 replied:
> The first sentence below is based on the fact that previously addCFIInstruction was used with OpNegateRAState, right?

Yes, but it's also based on that all other CFIs are added with `addCFIInstruction` and this one isn't, so it could look like a missing call to the addCFIInstruction function. But I will reword it a bit, so it's more clear.

paschalis-mpeis commented on bolt/include/bolt/Core/BinaryFunction.h:
```diff
@@ -1599,6 +1599,52 @@ class BinaryFunction {
+      I--;
+
+      switch (CFIOpcode) {
+      case dwarf::DW_CFA_GNU_window_save:
```
IMHO, the below looks more natural here:

```suggestion
      case dwarf::DW_CFA_AARCH64_negate_ra_state:
```

paschalis-mpeis commented on bolt/lib/Core/Exceptions.cpp:
```diff
@@ -632,8 +643,13 @@ bool CFIReaderWriter::fillCFIInfoFor(BinaryFunction &Function) const {
+        // mark that the instruction modifies the RA State. The actual state for
+        // instructions are worked out in MarkRAStates based on these
+        // annotations.
+        Function.setInstModifiesRAState(DW_CFA_GNU_window_save, Offset);
```
IMHO the below looks more natural:

```suggestion
        Function.setInstModifiesRAState(DW_CFA_AARCH64_negate_ra_state, Offset);
```

paschalis-mpeis commented on bolt/lib/Passes/InsertNegateRAStatePass.cpp:
```diff
@@ -0,0 +1,151 @@
+void InsertNegateRAState::runOnFunction(BinaryFunction &BF) {
+  BinaryContext &BC = BF.getBinaryContext();
+
+  if (BF.getState() == BinaryFunction::State::Empty) {
+    return;
+  }
```
nit: could drop braces

```suggestion
  if (BF.getState() == BinaryFunction::State::Empty)
    return;
```

paschalis-mpeis commented on bolt/lib/Passes/InsertNegateRAStatePass.cpp:
```diff
@@ -0,0 +1,151 @@
+        FoundAny = true;
+      }
+
+      if (BC.MIB->isPAuth(Inst)) {
```
nit: this could be an `else if` here, correct?

bgergely0 replied:
oops, it should actually just be an or: `if (BC.MIB->isPsign(Inst) || BC.MIB->isPauth(Inst))`

paschalis-mpeis commented on bolt/lib/Passes/InsertNegateRAStatePass.cpp:
```diff
@@ -0,0 +1,151 @@
+  BinaryContext &BC = BF.getBinaryContext();
+  bool FirstIter = true;
+  MCInst PrevInst;
+  for (auto BBIt = BF.begin(); BBIt != BF.end(); ++BBIt) {
+    BinaryBasicBlock &BB = *BBIt;
```
nit: Any reason not to iterate like below?
```cpp
for (BinaryBasicBlock &BB : BF)
```

paschalis-mpeis commented on bolt/lib/Passes/InsertNegateRAStatePass.cpp:
```diff
@@ -0,0 +1,151 @@
+  for (auto BBIt = BF.begin(); BBIt != BF.end(); ++BBIt) {
+    BinaryBasicBlock &BB = *BBIt;
+
+    for (auto It = BB.begin(); It != BB.end(); ++It) {
```
Would it be simpler to drop the `FirstIter` logic and start iterating from the 2nd element?
This will need another check that instructions are >=2 ..

This is personal preference, so I leave this up to you!

bgergely0 replied:
I think I did it this way, because I want to only skip the first iter on the first BB, so I cannot change the inner loop to only iterate from the second Inst. 
And then there is the case of the first two BBs having one Inst each, where I would need to set prev Inst some other way.
So overall I think this is easier, but I agree that it's not the cleanest code.

paschalis-mpeis commented on bolt/lib/Passes/MarkRAStates.cpp:
```diff
@@ -0,0 +1,122 @@
+// - NegateRAState flips the RA State,
+// - RememberState pushes the RA State to a stack,
+// - RestoreState pops the RA State from the stack.
+// These were saved as MCAnnotations on instructions they refer to at CFI
```
Thanks for adding a detailed description on the passes implementations. I find it very helpful!

Could you kindly reword a bit the first sentence here?
Also in the [PR comment](https://github.com/llvm/llvm-project/pull/120064#issue-2742064203) you may have a bit more info that you could amend here as well?

paschalis-mpeis commented on bolt/lib/Passes/MarkRAStates.cpp:
```diff
@@ -0,0 +1,122 @@
+          continue;
+        if (Annotation == MCPlus::MCAnnotation::kNegateState)
+          RAState = !RAState;
+        if (Annotation == MCPlus::MCAnnotation::kRememberState)
```
nit: Maybe using `else ifs` in the below 2 cases is more clear.

paschalis-mpeis commented on bolt/lib/Passes/MarkRAStates.cpp:
```diff
@@ -0,0 +1,122 @@
+  };
+
+  ParallelUtilities::runOnEachFunction(
+      BC, ParallelUtilities::SchedulingPolicy::SP_TRIVIAL, WorkFun, nullptr,
```
Is the policy chosen based on your discussion with @kbeyls  about costs?
Unsure what is more appropriate here. At least on theoretical time complexity you do a worst-case 2x linear pass over instructions.

bgergely0 replied:
No, I just copied this from some other Pass I believe, and forgot to make a proper decision. Do you have a suggestion to which one should I use?

paschalis-mpeis replied:
I believe it should be `SP_INST_LINEAR` instead (same for `InsertNegateRAStatePass.cpp`).

paschalis-mpeis replied:
@maksfb , @yota9 can you comment on this?

bgergely0 replied:
In [35306ef](https://github.com/llvm/llvm-project/pull/120064/commits/35306ef3cf370402717b8d1bd0ed8a829f0c4f09), I pushed some code related to the parallel execution, and wanted to fix this as well.

However, I've found that adding `SP_INST_LINEAR` breaks [`bolt/test/AArch64/unmarked-data.test`](https://github.com/llvm/llvm-project/blob/main/bolt/test/AArch64/unmarked-data.test). The test checks that no warnings are emitted, but we get this:

```
 // CHECK-NOT: BOLT-WARNING
               ^
<stdin>:7:1: note: found here
BOLT-WARNING: Running parallel work of 0 estimated cost, will switch to trivial scheduling.
```

Any tips on how to add `SP_INST_LINEAR` without breaking this?

paschalis-mpeis commented on bolt/lib/Core/MCPlusBuilder.cpp:
```diff
@@ -147,6 +147,92 @@ bool MCPlusBuilder::isTailCall(const MCInst &Inst) const {
+}
+
+bool MCPlusBuilder::hasNegateRAState(const MCInst &Inst) const {
+  if (hasAnnotation(Inst, MCAnnotation::kNegateState))
+    return true;
+  return false;
```
nit (could repeat for all has/is introduced methods below, excl. `isRAStateUnknown`)

```suggestion
  return hasAnnotation(Inst, MCAnnotation::kNegateState);
```

bgergely0 replied:
yupp, `isRAStateUnknown` would be a bit confusing that way

paschalis-mpeis commented on bolt/lib/Core/MCPlusBuilder.cpp:
```diff
@@ -147,6 +147,92 @@ bool MCPlusBuilder::isTailCall(const MCInst &Inst) const {
+}
+
+bool MCPlusBuilder::isRAStateUnknown(const MCInst &Inst) const {
+  if (hasAnnotation(Inst, MCAnnotation::kUnsigned) ||
+      hasAnnotation(Inst, MCAnnotation::kSigned) ||
+      hasAnnotation(Inst, MCAnnotation::kSigning) ||
+      hasAnnotation(Inst, MCAnnotation::kAuthenticating))
+    return false;
+  return true;
```
nit:  consider reusing your earlier functions, ie:
```cpp
return (isRAUnsigned(Inst) || isRASigned(Inst) || .. );
```

paschalis-mpeis reviewed (commented):
Hey Gergely, thanks a lot for your patch. It is very well written!

> The change has been tested on unittests binaries from the Chromium project, namely zlib_unittests and base_unittests.

It's great that you tested this externally with Chromium.
My high-level comment is about bolt testing. Any plans or thoughts about it?

You may also find some more comments below, mostly 'nits'. Thanks again!

whousemyname commented:
Hello bgergely0, This PR is really great!!
But when I used this PR, I instrumented my lib.so (aarch64) file, but the bolt crashed. It was related to the hash of the relocation information. I don’t know if it was because of the duplication of my lib.so file. Is it related to positioning information or PR?
The command executed is ` llvm-bolt.exe libGG.debug.so -instrument --instrumentation-file=GG.fdata -o .\libGG-debug.so.instrumented`
```
BOLT-INFO: shared object or position-independent executable detected
BOLT-INFO: Target architecture: aarch64
BOLT-INFO: BOLT version: dac8f161e868f57e34c5a12bc81c939a64322d15
BOLT-INFO: first alloc address is 0x0
BOLT-INFO: creating new program header table at address 0x4000000, offset 0x4000000
BOLT-WARNING: debug info will be stripped from the binary. Use -update-debug-sections to keep it.
BOLT-INFO: enabling relocation mode
BOLT-INFO: forcing -jump-tables=move for instrumentation
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: D:\\github-projects\\llvm-project\\build-forbolt\\bin\\llvm-bolt.exe D:\\projects\\apks\\v102\\forbolt-unity-relocs-1\\Output2\\libGameCore.debug.so -instrument --instrumentation-file=baldwinzhao_unity.fdata -o .\\libunity-debug.so.instrumented
Exception Code: 0x80000003
 #0 0x00007ff7dee0b653 std::_Tree_const_iterator<class std::_Tree_val<struct std::_Tree_simple_types<struct llvm::bolt::Relocation>>>::operator*(void) const C:\install_vs_buildtools\VC\Tools\MSVC\14.42.34433\include\xtree:182:0
 #1 0x00007ff7dff570f4 std::_Tree_const_iterator<class std::_Tree_val<struct std::_Tree_simple_types<struct llvm::bolt::Relocation>>>::operator->(void) const C:\install_vs_buildtools\VC\Tools\MSVC\14.42.34433\include\xtree:189:0
 #2 0x00007ff7dff5134c llvm::bolt::BinarySection::hash(class llvm::bolt::BinaryData const &, class std::map<class llvm::bolt::BinaryData const *, unsigned __int64, struct std::less<class llvm::bolt::BinaryData const *>, class std::allocator<struct std::pair<class llvm::bolt::BinaryData const *const, unsigned __int64>>> &) const D:\github-projects\llvm-project\bolt\lib\Core\BinarySection.cpp:58:0
 #3 0x00007ff7dffc4bf4 llvm::bolt::BinarySection::hash(class llvm::bolt::BinaryData const &) const D:\github-projects\llvm-project\bolt\include\bolt\Core\BinarySection.h:429:0
 #4 0x00007ff7dff64c47 llvm::bolt::BinaryContext::generateSymbolHashes(void) D:\github-projects\llvm-project\bolt\lib\Core\BinaryContext.cpp:1186:0
 #5 0x00007ff7dff6d11a llvm::bolt::BinaryContext::postProcessSymbolTable(void) D:\github-projects\llvm-project\bolt\lib\Core\BinaryContext.cpp:1414:0
 #6 0x00007ff7ded4a796 llvm::bolt::RewriteInstance::buildFunctionsCFG(void) D:\github-projects\llvm-project\bolt\lib\Rewrite\RewriteInstance.cpp:3431:0
 #7 0x00007ff7ded3936e llvm::bolt::RewriteInstance::run(void) D:\github-projects\llvm-project\bolt\lib\Rewrite\RewriteInstance.cpp:688:0
 #8 0x00007ff7dd07eb8b main D:\github-projects\llvm-project\bolt\tools\driver\llvm-bolt.cpp:268:0
 #9 0x00007ff7e27221c9 invoke_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:79:0
#10 0x00007ff7e27220b2 __scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288:0
#11 0x00007ff7e2721f6e __scrt_common_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:331:0
#12 0x00007ff7e272225e mainCRTStartup D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_main.cpp:17:0
#13 0x00007ffac8cd259d (C:\WINDOWS\System32\KERNEL32.DLL+0x1259d)
#14 0x00007ffaca30af38 (C:\WINDOWS\SYSTEM32\ntdll.dll+0x5af38)
```

bgergely0 commented:
@whousemyname Hi! Pretty sure this code has nothing to do with the BinarySection::hash function. Until this PR is merged, and considered "stable", I suggest to only use executables compiled with `-mbranch-protection=none` with BOLT, and use the code from the main branch.

bgergely0 commented:
Thanks for the detailed look at it @paschalis-mpeis ! 
I addressed most of your comments. The scheduling policy is a missing bit.

Testing:
We definately should add some tests next to the code. Problem is that the code needs a faily complex test (e.g. it needs to rely on an unwinder). So the smallest possible target we could (re)use is tests from the libunwind subproject. [This file for example](https://github.com/llvm/llvm-project/blob/main/libunwind/test/libunwind_01.pass.cpp).

bgergely0 commented:
I will have a small improvement up in a few days to add some error recovery capability to the [asserts here](https://github.com/llvm/llvm-project/blob/fee74748ef374b77d22710e45d1ef09c432f03c8/bolt/lib/Passes/MarkRAStates.cpp#L74-L85). 
In MarkRAStates, I assume that functions start in an unsigned RA state, and that's why these asserts are in place. I have encountered one situation with handwritten assembly, where this is not be the case: a pair of paciasp and autiasp insts were split into two functions. To not fail on these asserts, I wrote [this hacky workaround](https://github.com/llvm/llvm-project/blob/fee74748ef374b77d22710e45d1ef09c432f03c8/bolt/lib/Passes/MarkRAStates.cpp#L46-L62) in MarkRAStates.cpp. 
But after submitting, I now realize that this situation is analogous to the normal function-splitting done during optimizations (so a function is split, and the paciasp and autiasp end up in different functions). So some level or error-recovery from the asserts is possible, which would make the code cleaner, and less likely to fail on the asserts.

paschalis-mpeis commented on bolt/include/bolt/Passes/InsertNegateRAStatePass.h:
```diff
@@ -16,7 +27,15 @@ class InsertNegateRAState : public BinaryFunctionPass {
+private:
+  /// Loops over all instructions and adds OpNegateRAState CFI
+  /// after any pointer signing or authenticating instructions.
+  /// Returns true, if any OpNegateRAState CFIs were added.
```
nit: `\return` true if ..

bgergely0 replied:
I don't see that notation anywhere in BOLT, so I'm not sure adding here improves much.

paschalis-mpeis replied:
This is more on [doxygen](https://llvm.org/docs/CodingStandards.html#doxygen-use-in-documentation-comments) guidelines. But I also see this in a few places, eg [findAttributeInfo](https://github.com/llvm/llvm-project/blob/6b1db79887df19bc8e8c946108966aa6021c8b87/bolt/include/bolt/Core/DebugData.h#L61).

This is a nit anyway, so if you remember add it on a follow-up change, when we have some update on tests or the scheduling policy, etc.

paschalis-mpeis commented on bolt/lib/Core/Exceptions.cpp:
```diff
@@ -640,16 +640,16 @@ bool CFIReaderWriter::fillCFIInfoFor(BinaryFunction &Function) const {
+      // same id but mean different things. The latter is used in AArch64.
       if (Function.getBinaryContext().isAArch64()) {
-        // Not adding OpNegateRAState since the location they are needed
+        // The location OpNegateRAState CFIs are needed
```
I think you could drop 'are' here?

// The location OpNegateRAState CFIs ~~are~~ needed

bgergely0 replied:
Grammar-wise I believe it is needed.

paschalis-mpeis commented on bolt/include/bolt/Core/MCPlus.h:
```diff
@@ -72,7 +72,14 @@ class MCAnnotation {
     kSize,                /// Size of the instruction.
     kDynamicBranch,       /// Jit instruction patched at runtime.
-    kGeneric              /// First generic annotation.
+    kSigning,             /// Inst is a signing instruction (paciasp, etc.)
```
nit: could end comments here with dots.

paschalis-mpeis commented on bolt/lib/Core/MCPlusBuilder.cpp:
```diff
@@ -147,6 +147,76 @@ bool MCPlusBuilder::isTailCall(const MCInst &Inst) const {
+}
+
+bool MCPlusBuilder::isRAStateUnknown(const MCInst &Inst) const {
+  if (isRAUnsigned(Inst) || isRASigned(Inst) || isRASigning(Inst) ||
+      isAuthenticating(Inst))
+    return false;
+  return true;
```
nit:
```suggestion
  return (isRAUnsigned(Inst) || isRASigned(Inst) || isRASigning(Inst) ||
      isAuthenticating(Inst));
```

paschalis-mpeis commented on bolt/lib/Passes/MarkRAStates.cpp:
```diff
@@ -0,0 +1,122 @@
+      if (BC.MIB->isPSign(Inst)) {
+        assert(!RAState && "Signed RA State before PSign");
+        BC.MIB->setRASigning(Inst);
+
```
nit: could drop empty line
```suggestion
```

paschalis-mpeis reviewed (commented):
Thanks for addressing my comments @bgergely0!

I've added a few more nits. Let's wait on the scheduling policy and feel free to reach out regarding tests.

paschalis-mpeis commented:
@whousemyname you could double-check by trying the first commit before this PR (`3063eaf85~`) that it still has the crash you mention.

whousemyname commented:
> @whousemyname you could double-check by trying the first commit before this PR (`3063eaf85~`) that it still has the crash you mention.

Thank you for your suggestion. The cause of this crash should have existed before. The llvm-bolt I built on mac did not encounter this problem, but the llvm-bolt built using msvc on Windows encountered this hash crash problem.

paschalis-mpeis commented:
> The change has been tested on unittests binaries from the Chromium project, namely zlib_unittests and base_unittests.

Regarding some end-to-end testing for PAuth, like with zlib in Chromium, we could maybe add some small artifact to [rafaelauler/bolt-tests](https://github.com/rafaelauler/bolt-tests)
(cc: @ilinpv)

ostannard commented:
I'm not familiar with the internals of BOLT, but this looks like a reasonable strategy for handling return address signing. The only comments I have should be taken as suggestions for future patches, and probably don't need to be handled before this can be committed.

When compiling with `-march=armv8.3-a` (or later, or a later `-mcpu` option), there are new instructions which get used, instead of the NOP instructions this checks for. I think we always use the NOP instructions (PACIxSP) in the prologue, but we emit the new v8.3-A RETAx instructions for returns.

The compiler can also emit the 'B' character into the CIE augmentation string to signal that the B key (instead of the default A) was used for return address signing. This can be enabled in the compiler with `-mbranch-protection=pac-ret+b-key` or `-fptrauth-returns`. I don't think any platforms mix the two keys for return address signing in one process, but the unwinder does use the augmentation string to decide which authentication instruction to use, so if we create new CFIs then we'll need to make sure that gets set correctly.

Armv9.5-A added a new extension to pointer authentication, FEAT_PAUTH_LR, which adds the PC address of the signing instruction to the signature. There is a new unwind opcode for this, `DW_CFA_AARCH64_negate_ra_state_with_pc`, which toggles the signing state and captures the code address to be used for authentication during unwinding. There are a few different forms of the epilogue, depending on both the architecture version and the length of the function, and BOLT will need to be able to adjust these to get the correct address. This also adds the extra constraint that there can only be one signing instruction which dominates each authenticating instruction, I don't know if any BOLT optimisations are likely to break that or not.

Heart-tide commented:
Hi @bgergely0 . Thank you for your PR! It works well in most scenarios. However, while testing on `vmlinux`, I occasionally encountered a segmentation fault at `InsertNegateRAStatePass.cpp:81`.

Observations:
- The crash occurs at `PrevInst = Inst` after `BF.addCFIInstruction()` is called.
- When I modified the code to use a local copy of `*It` (instead of a reference `MCInst& Inst`), the crash no longer reproduced.

Hypothesis:
The `vector::insert` operation within `addCFIInstruction()` may invalidate existing iterators/references (per C++ iterator invalidation rules). This could turn `Inst` into a dangling reference if the vector reallocates, leading to UB when accessed later.

Since the issue is non-deterministic (depends on reallocation), it might explain the intermittent crashes. Would you please review the iterator/reference handling around the CFI insertion point? Thank you for your time!

bgergely0 commented:
Hi @Heart-tide, thanks for reporting this! Your observations seems correct: the `It = BF.addCFIInstruction()` updates the iterator, but the `Inst` variable is still a reference to the original one. I will fix this soon. Let me know if you find any other issues with the patch!

bgergely0 committed:
[BOLT] Add unit tests for negate_ra_state cfi handling

- also add match_dwarf.py, a tool used by the unit tests.

bgergely0 committed:
[BOLT][AArch64] Fix which PSign and PAuth variants are used (#120064)

- only the ones operating on LR should be marked
    with .cfi_cfi_negate_ra_state
- added support for fused PtrAuth and Ret instructions,
    e.g. RETAA.

bgergely0 commented:
force push: rebased this to the current main, because it was outdated

paschalis-mpeis commented on bolt/test/AArch64/incorrect-negate-ra-state.s:
```diff
@@ -0,0 +1,41 @@
+# RUN: llvm-mc -filetype=obj -triple aarch64-unknown-unknown %s -o %t.o
```
Can you add here a brief description of the test, explaining why this case is wrong.

paschalis-mpeis commented on bolt/test/AArch64/incorrect-negate-ra-state.s:
```diff
@@ -0,0 +1,41 @@
+# CHECK: BOLT-INFO: inconsistent RAStates in function foo
+
+# check that foo got Ignored, so it's not in the new .text section
+# llvm-objdump %t.exe -d -j .text > %t.exe.dump
+# RUN: not grep "<foo>:" %t.exe.dump
```
I believe llvm-objdump was meant to be a RUN line and applied to `%t.exec.bolt` instead.

Currently it's not failing as`%t.exe.dump` is never created.

paschalis-mpeis commented on bolt/test/AArch64/incorrect-negate-ra-state.s:
```diff
@@ -0,0 +1,41 @@
+  .type _start, %function
+_start:
+  b foo
+
```
nit: add the usual comment:

```
# Force relocation mode.
```

However, I don't think relocation mode is needed for this test, so you could drop it instead.

paschalis-mpeis commented on bolt/test/match_dwarf.py:
```diff
@@ -0,0 +1,114 @@
+# (= the output from running llvm-objdump --dwarf=frames),
+# by address to function names (which are parsed from a normal objdump).
+# The script is used for checking if .cfi_negate_ra_state CFIs
+# are generated by BOLT the same way they are generated by LLVM.
```
Maybe expand with an extra 1-2 lines how this is achieved?

eg, outputting negate offsets – and say why that is a good metric for comparisons.

paschalis-mpeis commented on bolt/test/match_dwarf.py:
```diff
@@ -0,0 +1,114 @@
+        negate_offsets = []
+        loc = 0
+        # TODO: make sure this is not printed in hex
+        re_advloc = f"DW_CFA_advance_loc: (\d+)"
```
```suggestion
        re_advloc = f"DW_CFA_advance_loc: (\\d+)"
```

Otherwise I get:
> SyntaxWarning: invalid escape sequence '\d'

bgergely0 replied:
Quite surprising to see that this works for me locally, and the CI tests also pass.

paschalis-mpeis commented on bolt/test/match_dwarf.py:
```diff
@@ -0,0 +1,114 @@
+
+
+def parse_objdump(objdump):
+    """
+    Parse and return address-to-name dictionary from objdump file
+    """
```
nit: consider adding a quick example of what you're trying to match:
```
000123abc <foo>:
```

paschalis-mpeis commented on bolt/test/match_dwarf.py:
```diff
@@ -0,0 +1,114 @@
+
+
+def parse_dwarf(dwarfdump, addr_name_dict):
+    """
+    Parse dwarf dump, and match names to blocks using the dict from the objdump.
+    Return a list of NameDwarfPairs.
+    """
```
Same here with:
```
000123 000456 000789 FDE cie=000000  pc=0123abc...0456def
```

then clarify that you first extract the `pc` and later store the DWARF info for the function whose addressed matched earlier.

paschalis-mpeis commented on bolt/test/match_dwarf.py:
```diff
@@ -0,0 +1,114 @@
+        functions = []
+        for line in dw.readlines():
+            match = re_address_line.match(line)
+            if not match:
+                if len(functions) > 0:
+                    functions[-1].append(line)
+                continue
```
This keeps adding lines till the end of the dump file (at least when using `-d`). You could probably end when you reach:
> Disassembly of section .text:

Or alternatively at `ZERO terminator`, though that would require more changes to the logic.

BTW consider adding another test with two functions containing dwarf data. It might help you answer the above and also cover a slightly more complex scenario.

paschalis-mpeis commented on bolt/test/match_dwarf.py:
```diff
@@ -0,0 +1,114 @@
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("objdump", help="Objdump file")
```
Do we really need this objdump file? I think we can use `parse_objdump` on `args.dwarfdump` too, and drop the extra argument, and the RUN line that generates `%t.exe.dump` in the lit test.

Given that's done, you could consider renaming:
- `parse_objdump`to something like `extract_symbols` or `extract_functions`
- `parse_dwarf` to `associate_dwarf_data`

(just quick suggetions – feel free to choose better names)

bgergely0 replied:
Will change the names, addressed the combined disassembly in prev comment.

paschalis-mpeis commented on bolt/test/match_dwarf.py:
```diff
@@ -0,0 +1,114 @@
+    addr_name_dict = parse_objdump(args.objdump)
+    functions = parse_dwarf(args.dwarfdump, addr_name_dict)
+
+    for f in functions:
```
Are there any plans to support multiple functions?
If not, can use `--disassemble-symbols=SYMBOL` in lit and drop the loop here.

bgergely0 replied:
Good point, I think the current way the script looks (I mean it only outputs the offsets for one specific function, which we pipe to diff), it is better to only support one function at a time, and simplify the disassembly with that flag.

paschalis-mpeis commented on bolt/test/AArch64/simple-negate-ra-state.s:
```diff
@@ -0,0 +1,42 @@
+# RUN: llvm-mc -filetype=obj -triple aarch64-unknown-unknown %s -o %t.o
```
Please add a short description to the test.

nit: you can drop 'simple-' from the title.
Maybe name both tests like `negate-ra-state.s` and `negate-ra-state-incorrect.s`, so they appear next to each other.

paschalis-mpeis commented on bolt/test/AArch64/simple-negate-ra-state.s:
```diff
@@ -0,0 +1,42 @@
+# RUN: llvm-mc -filetype=obj -triple aarch64-unknown-unknown %s -o %t.o
+# RUN: %clang %cflags  %t.o -o %t.exe -Wl,-q
+
+# RUN: llvm-objdump %t.exe -d > %t.exe.dump
```
I don't think we need `.dump-dwarf` files – see more at `match_dwarf.py`

paschalis-mpeis commented on bolt/test/AArch64/simple-negate-ra-state.s:
```diff
@@ -0,0 +1,42 @@
+# RUN: %clang %cflags  %t.o -o %t.exe -Wl,-q
+
+# RUN: llvm-objdump %t.exe -d > %t.exe.dump
+# RUN: llvm-objdump --dwarf=frames %t.exe -D > %t.exe.dump-dwarf
```
Do we need `-disassemble-all` here, or would `-d` be enough?
Or better yet, since python processes a single function, can use:
`--disassemble-symbols=foo`

Note: I noticed your bolted `.dump-dwarf` doesn't include any `.text`.
Still, you could use a single file that contains just the `.text` of foo.
What do you think?

bgergely0 replied:
I think it is better to have the dwarf info and the normal disassembly in separate files, but disassembling one foo is definately better than using the `-d` flag.

paschalis-mpeis commented on bolt/test/AArch64/simple-negate-ra-state.s:
```diff
@@ -0,0 +1,42 @@
+
+# RUN: llvm-objdump %t.exe -d > %t.exe.dump
+# RUN: llvm-objdump --dwarf=frames %t.exe -D > %t.exe.dump-dwarf
+# RUN: match-dwarf %t.exe.dump %t.exe.dump-dwarf foo > orig.txt
```
re `orig.txt`, consider using `%t`, maybe something like `%t.match-dwarf.orig.txt`

paschalis-mpeis commented on bolt/test/AArch64/simple-negate-ra-state.s:
```diff
@@ -0,0 +1,42 @@
+
+# RUN: llvm-objdump %t.exe.bolt -d > %t.exe.bolt.dump
+# RUN: llvm-objdump --dwarf=frames %t.exe.bolt  > %t.exe.bolt.dump-dwarf
+# RUN: match-dwarf %t.exe.bolt.dump %t.exe.bolt.dump-dwarf foo > bolted.txt
```
Same as with `orig.txt`

paschalis-mpeis commented on bolt/test/match_dwarf.py:
```diff
@@ -0,0 +1,114 @@
+        print(self.name)
+        print(self.body)
+
+    def parse_negates(self):
```
Consider adding a short description – you are summing the advance_loc entries and appending only when a negate_ra_state directive is found

paschalis-mpeis reviewed (commented):
Thanks Gergely for adding the tests!


Currently, these CFIs are processed without requiring any flag, ie this works out of the box:

>  llvm-bolt %t.exe -o %t.exe.bolt

Since this is a new feature, do you think it might be worth guarding it with a flag to make it more explicit to users? For example, fail with some advice when the flag is not present, otherwise process Negate-RA CFIs.

Regardless, I think it would be helpful to emit a BOLT-INFO stat summarizing what was done. So users are aware that this feature was used and applied to N places.

Please find a few more comments below. Thanks again. Great work! :)

bgergely0 commented:
@paschalis-mpeis 

> Since this is a new feature, do you think it might be worth guarding it with a flag to make it more explicit to users? For example, fail with some advice when the flag is not present, otherwise process Negate-RA CFIs.

Good point, will do. 

> Regardless, I think it would be helpful to emit a BOLT-INFO stat summarizing what was done. So users are aware that this feature was used and applied to N places.

I am not sure if this would be very useful for users. 

We expect BOLT users to optimize big, complex programs (compilers, database engines, etc), so the generated amount would be huge. Almost all functions are affected. 

Because of this, it would be way easier for users to switch off the flag, and use a binary with `-mbranch-protection=none` if they face issues, than to debug which exact function's missing/extra RA state negation causes the unwinder to fail.

atrosinenko commented on bolt/lib/Passes/MarkRAStates.cpp:
```diff
@@ -0,0 +1,133 @@
+  for (BinaryBasicBlock &BB : BF) {
+    for (auto It = BB.begin(); It != BB.end(); ++It) {
+      MCInst &Inst = *It;
+      if ((BC.MIB->isPSign(Inst) || BC.MIB->isPAuth(Inst)) &&
+          !BC.MIB->hasNegateRAState(Inst)) {
+        // no .cfi_negate_ra_state attached to signing or authenticating instr
+        // means, that this is a function with handwritten assembly, which might
+        // not respect Clang's conventions (e.g. tailcalls are always
+        // authenticated, so functions always start with unsigned RAState when
+        // working with compiler-generated code)
+        BF.setIgnored();
+        BC.outs() << "BOLT-INFO: inconsistent RAStates in function "
+                  << BF.getPrintName() << "\n";
+        return;
+      }
```
If I understand correctly, this check would reject code like the following:
```
  autiasp x0, x1
  blr     x0
```
which probably doesn't violate anything (even though `blra x0, x1` would likely be emitted by Clang instead). On the other hand, `autia1716` is encoded as `hint`, so this could even be useful in real-world code, as it is compatible with Armv8.2-a:
```
  autia1716
  blr       x17
```

I guess, this pass could check not "is this an auth/sign instruction", but "is this an auth/sign instruction **operating on LR**". Possibly, target-specific hooks like [`getAuthenticatedReg`](https://github.com/llvm/llvm-project/blob/main/bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp#L199) and `getTrustedLiveInRegs` (to guess "the LR" without referring to `AArch64::LR` from target-independent code) could be reused.

paschalis-mpeis commented:
>
> > Since this is a new feature, do you think it might be worth guarding it with a flag to make it more explicit to users? For example, fail with some advice when the flag is not present, otherwise process Negate-RA CFIs.
> 
> Good point, will do.
> 
> >  Regardless, I think it would be helpful to emit a BOLT-INFO stat summarizing what was done. So users are aware that this feature was used and applied to N places.
> 
> I am not sure if this would be very useful for users.
>

My reasoning was to output information so users are aware of what's happening under the hood.
Guarding this with a flag is even better (and my preference).

---

> Possibly, target-specific hooks like [getAuthenticatedReg](https://github.com/llvm/llvm-project/blob/main/bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp#L199) and getTrustedLiveInRegs (to guess "the LR" without referring to AArch64::LR from target-independent code) could be reused.

@atrosinenko sounds like a good idea.

bgergely0 commented:
Note to myself/reviewers: #121559 has changed how `.cfi_negate_ra_state`s are generated for `paciasp`/`autiasp`, which brakes my assumptions about the input binary. The change there needs to be reflected in this PR as well.

paschalis-mpeis reviewed (commented):
Hey Gergely,

Thank you so much for your work!
I haven't finished reviewing as the patch is big, but I figured it might be helpful to share what I have so far.

Please have a look.

bgergely0 commented on bolt/lib/Rewrite/BinaryPassManager.cpp:
```diff
@@ -273,6 +275,11 @@ static cl::opt<bool> ShortenInstructions("shorten-instructions",
                                          cl::desc("shorten instructions"),
                                          cl::init(true),
                                          cl::cat(BoltOptCategory));
+
+cl::opt<bool> AllowPacret(
+    "allow-experimental-pacret",
+    cl::desc("Enable processing binaries with pac-ret (experimental)"),
+    cl::cat(BoltOptCategory));
```
That's quite surprising, I've built this branch on several machines, plus the CI build+tests also passed. I will look at this.

bgergely0 replied:
Still couldn't reproduce this, but I have moved it to `CommandLineOpts.cpp` just in case.

paschalis-mpeis commented on bolt/test/AArch64/negate-ra-state.s:
```diff
@@ -0,0 +1,42 @@
+# RUN: %clang %cflags  %t.o -o %t.exe -Wl,-q
+
+# RUN: llvm-objdump %t.exe -d > %t.exe.dump
+# RUN: llvm-objdump --dwarf=frames %t.exe -D > %t.exe.dump-dwarf
```
nit: can you please drop `-D` from here?

bgergely0 commented on bolt/lib/Core/BinaryFunction.cpp:
```diff
@@ -2769,10 +2770,12 @@ struct CFISnapshot {
-        llvm_unreachable("BOLT-ERROR: binaries using pac-ret hardening (e.g. "
-                         "as produced by '-mbranch-protection=pac-ret') are "
-                         "currently not supported by BOLT.");
+      if (!(opts::BinaryAnalysisMode || opts::HeatmapMode ||
+            opts::AllowPacret)) {
+        llvm_unreachable(
+            "BOLT-ERROR: support for binaries using pac-ret hardening (e.g. as "
+            "produced by '-mbranch-protection=pac-ret') is experimental\n"
+            "BOLT-ERROR: set --allow-experimental-pacret to allow processing");
```
To not fail on binaries compiled with `-mbranch-protection=none` but on a system where the C runtime objects (e.g. `crtbegin.o`) were compiled with `pac-ret`, we cannot check for all `.cfi_negate_ra_state`. Doing so would also break a lot of unittests on the aarch64 runner.

We have two options:
- find all C runtime function names (including function names from different libc implementations), and ignore `.cfi_negate_ra_state`s in them OR
- move the check later (in time) to the point where processing CFIs previously crashed for OpNegateRAState. 

I've implemented the second option here, so that explains why there is no crash for the unittests. Only those binaries crash where at least one function triggers `unwindCFIState`. This approach is a bit "probabilistic", the other could be more robust, but that requires manual check for function names.

This illustrates that adding the `--allow-experimental-pacret` flag introduces extra complexity which we would not have without it.

paschalis-mpeis reviewed (commented):
Hey Gergely,

Thanks for addressing the review – I think the patch is now mature enough to come out of draft mode soon!

We discussed this offline and it seems a [bolt/docs](https://github.com/llvm/llvm-project/tree/main/bolt/docs) entry explaining this great work and why it matters, along with a 2-3 more tests (eg, code that does actual reordering/splitting/..) would make this complete. Some end-to-end coverage could also come to [bolt-tests](https://github.com/rafaelauler/bolt-tests).

We've seen distros start enabling this feature – the [many failed tests](https://github.com/llvm/llvm-project/pull/136664#issuecomment-2835537481) on #136664 suggest this will only become more important.

On [comment for #120064](https://github.com/llvm/llvm-project/pull/120064#issuecomment-2821064959): thanks for escalating  – are we looking for a possible revert?


---

On this review itself, just a couple of nits.

bgergely0 committed:
[BOLT] Added docs/PacRetDesign.md

bgergely0 committed:
[BOLT] Basic exception unwinding test

bgergely0 commented:
Force push: I had to rebase to main, so I have the change from #137795 for C/C++ tests. See [this earlier comment](https://github.com/llvm/llvm-project/pull/120064#issuecomment-2821064959).

bgergely0 commented:
Force push: an implementation of `getSignedReg` has been merged into main, so my previous rebase had both with a slightly different signature.

bgergely0 committed:
[BOLT] Add OpNegateRAState to printCFI

bgergely0 committed:
[BOLT] Improve function splitting at OpNegateRAState handling

- Previous version used the Layout API which should not be used.
- This version iterates on Fragments to have the same effect.
- NegateRAState is added to the first *non-empty* BB of the cold fragment.

bgergely0 committed:
[BOLT] Bugfix: CFIs can be placed before the first Instruction

- This caused a crash when trying to Annotate RAState-changing CFIs
    (RememberState, RestoreState, NegateRAState).
- The fix introduces an InitialRAState for each BinaryFunction.
- If we have a NegateRAState before the first Instr, we set that to
  True.
- In MarkRAStates, we push the InitialRAState to the RAStateStack: as we
  may have omitted the RememberState at the function start, its RestoreState
  pair would try to pop an empty stack otherwise.

bgergely0 committed:
[BOLT] Add function-splitting test

bgergely0 committed:
[BOLT] Improve warnings in MarkRAStates

- Previously, the same warning was printed in
  3 different cases. Changed warnings to be more specific.

bgergely0 committed:
[BOLT] Improve negate-ra-state-incorrect test

- check all 3 possibilities that MarkRAStates warns about,
- simplify the asm functions in the test,
- use FileCheck instead of grep on the disassembly.

bgergely0 committed:
[BOLT] Do run negate-ra-state rewriting passes on all functions

- if a BinaryFunction did not contain any .cfi_negate_ra_state,
  we can skip MarkRAStates and InsertNegateRAStatePass. This can happen
  with handwritten assembly, or if the binary was compiled in a way that
  drops DWARF tables.

bgergely0 committed:
[BOLT] Introduce --disallow-pacret flag

bgergely0 committed:
[BOLT][NFC] Review nits

- remove unused imports
- fix clangd warnings
- rename functions
- improve comments
- reformat lib/Core/Exceptions.cpp

bgergely0 committed:
[BOLT] Refactor and improve InsertNegateRAStatePass

- separate function splitting code into its own function
- change the iteration looking for state-switches between consecutive
  instructions to only look for changes *inside* FunctionFragments,
  and skip borders between them.

bgergely0 committed:
[BOLT] Add function splitting lit test without runtime dependency

- Test uses lit and FileCheck to check if OpNegateRAState CFIs are
  generated in the correct location for function-splitting.
- Also fix a bug related to function-splitting.

bgergely0 committed:
[BOLT][NFC] Simplify RAState tracking

- Remove 'Authenticating' and 'Signing' MCAnnotations.
- The same logic can be done using only 'Signed' and 'Unsigned' annotations.
- To check if an instruction is signing or authenticating, we can use
  the PSignOnLR, and PAuthOnLR functions.

bgergely0 committed:
[BOLT] Remove unnecessary script, and rewrite unit test using it

- the checks that match_dwarf.py helped with can also be done using BOLT
  logs.
- rewrote the negate-ra-state.s test to only rely on logs to verify
  correct OpNegateRAState CFI generation.

bgergely0 committed:
[BOLT] Add negate-ra-state-reorder test

- Test checks that that reordering BBs change the number of
  negate-ra-state CFIs, and their locations.

bgergely0 committed:
[BOLT] Print stats in MarkRAStates, InsertNegateRAState

- print statistics on the output about function getting Ignored in
  MarkRAStates, and on rewriting OpNegateRAState in
  InsertNegateRAStatePass.
- improve error messages in MarkRAStates

bgergely0 commented:
I've added tests, and created a doc to aid review and for future reference. If it's not something that should be in the patch, I can remove later.

@maksfb you mentioned using a MetaDataRewriter for this on the RFC post, instead of using the two passes. 
I looked into that, and I believe it can work with MarkRAStates becoming the `postCFGInitializer` and InsertNegateRAState becoming the `preEmitFinalizer`.  This will still need a bit of testing, but I marked the PR ready for review, as the contents of that will be the same as the contents of the two passes now.

bgergely0 committed:
[BOLT] Review nits

- fix spelling in MarkRAStates
- improve negate-ra-state.s test
- improve pacret-split-funcs.s test

bgergely0 commented:
Improved the docs (force push), and improved some tests and spelling.

bgergely0 committed:
[BOLT] Fix multithreading in #120064

- use SkipPredicates,
- use atomic counters.

paschalis-mpeis reviewed (commented):
Hey Gergely,

Nice refactor. Thanks for the docs, and the work!

Two unresolved topics remain: complexities ([link](https://github.com/llvm/llvm-project/pull/120064#discussion_r1916593521)), and the flag ([link](https://github.com/llvm/llvm-project/pull/120064#discussion_r2088438363)).

For the flag, we went with a negative `disallow-pacret`, which is off by default.

Q1) Do we anticipate additional modes, eg refuse to emit a binary on CFI inconsistencies? If so, we could rename to something like:
`-update-branch-protection={yes|strict|no}`

Q2) Running this by default increases the risk of breakage. I recommend we proceed, monitor, and revert if needed.

Any objections, or other concerns? It'd be great if Meta could take a look; and we can discuss tomorrow during Office Hours.

bgergely0 commented on bolt/docs/PacRetDesign.md:
```diff
@@ -0,0 +1,154 @@
+
+This bit indicates to the unwinder whether the current return address is signed or not (hence the name). The unwinder uses this information to authenticate the pointer, and remove the Pointer Authentication Code (PAC) bits. Incorrect negate-ra-state placement can lead to the unwinder trying to authenticate an unsigned pointer (which segfaults), or skipping authenticating a signed pointer, and try ...
+
+(Note: not *all* unwinders do this. Some use the `xpac` instruction to strip the PAC bits without authenticating the pointer. This is incorrect, as it allows control-flow modification in the case of unwinding.)
```
It is an _incorrect_ implementation of control flow integrity, as it leaves a gap (unwinding) in the integrity enforcing. So not incorrect as "possibly segfaulting". Could be called bad practise as well.

bgergely0 committed:
[BOLT] Address review

- improve docs/PacRetDesign.md
- improve pacret-split-funcs test

bgergely0 committed:
[BOLT] Change scheduling policy to SP_INT_LINEAR

- count the total applicable functions, and only run the passes if it's
  more than 0. This way we do not get a warning on the scheduling policy
  change.

paschalis-mpeis reviewed (approved):
Looks good. There is no other ask on my end in this initial patch for PAC. Further improvements and features will follow-up as needed.

As discussed last night in Office Hours, Meta will have a look too, so please wait for their input.

Thanks again, Gergely!

bgergely0 committed:
[BOLT] Rename pac-ret feature flag

- new name is update-branch-protection, which can be extended later to
  include a "strict mode", and to cover BTI rewriting.

bgergely0 commented:
Hi @maksfb! Just a heads-up: we plan to merge this early next week.

Please take a look, if you have the time. Otherwise we can merge, and keep an eye on issues. 
Additional passes are only running in the smallest scope they need: on aarch64, when the input has pac-ret, so the chance of breaking other things is minimal.

maksfb reviewed (commented):
Thanks again for adding the support for PAC-RET! Sorry for delayed review.

What is the policy for passes that insert instructions, is it their responsibility for annotating the state or is it something `InsertNegateRAStatePass` can handle?

bgergely0 committed:
Update InsertNegateRAStatePass

- fix spelling

Co-authored-by: Maksim Panchenko <maks@meta.com>

bgergely0 committed:
[BOLT] Review changes

- rename kSigned/kUnsigned to kRASigned/kRAUnsigned
- const& in MarkRAStates
- spelling
- change flag to hidden

bgergely0 committed:
[BOLT] Remove workaround

Remove the workaround for the issue that was fixed in #160263.

bgergely0 commented:
force push: formatting

bgergely0 commented:
Thanks for the reviews @maksfb !

> What is the policy for passes that insert instructions, is it their responsibility for annotating the state or is it something InsertNegateRAStatePass can handle?

Adding book-keeping responsibility to other passes would be a burden, so `InsertNegateRAStatePass` tries to guess these using the `inferUnknownStates` function. This just takes the RAState of the prev. instruction, unless that was a signing/authenticating inst, in which case takes the opposite.

This was written under the assumption that new instructions are mostly branches at the ends of BasicBlocks resulting from the reordered BBs. But now I realize that Stubs inserted in LongJmp are tricky, because two calls to the same target can share a Stub, but they might come from ranges with opposing RAStates, in which case the Stub's state is not clear.

Not sure of the solution here, we might have to limit reusing Stubs in such cases.

Another improvement could be to check if we are inserting after an instruction that has no fallthrough case (e.g. branch, return), meaning the next instruction is not after it in the control flow, and cannot inherit its RAState.

bgergely0 commented:
Hi @maksfb @paschalis-mpeis 

I looked at how unwinding works _from Stubs_, and opened this issue: https://github.com/llvm/llvm-project/issues/160989.

TLDR: while executing the Stubs, we cannot unwind as it is, so PAC-specific unwinding will also not work. In fact, to match the unwind info, it makes sense to have the same RAState in the stub, as before it.

For other new instructions, I think we can perfectly match the RAState (I don't expect other places where a Block is jumped over, so it is not part of the CFG of the function it is placed in).

bgergely0 committed:
[BOLT] Update bolt/docs/PacRetDesign.md

bgergely0 committed:
Merge branch 'main' into negate-ra-state-support-v2

bgergely0 commented:
@paschalis-mpeis @maksfb Ping

I have a wip change to make the `inferUnknownStates` more precise, but I don't want to push more changes to this PR, as it is already quite big. I also cannot add the followup change as a stacked PR, because this is a branch in my own fork. 

Do you have an issue with merging this as it is, and adding more changes on top afterwards?

paschalis-mpeis reviewed (approved):
Hey Gergely,

Thanks for updates. Agreed, lets keep the patch as is and merged it.

I reviewed the latest set of patches (11e897deffa6aac23ac533cf83697a9bb8bf8456 - fd3642ad858e98c87b8d5c39153bc2532b90680c).

Looks good. Feel free to merge with this initial patch by EOD, or tomorrow, given @maksfb has no further comments.

bgergely0 committed:
Update bolt/docs/PacRetDesign.md

spelling

Co-authored-by: Paschalis Mpeis <paschalis.mpeis@gmail.com>
